File: .claude/skills/mcp-api-integration/figma_client.py
"""Figma REST client. Python 3.9+, standard library only.

    FIGMA_TOKEN=<personal access token> python figma_client.py [figma-url]

An OAuth access token works too - pass token_type="oauth".
"""

import json
import os
import sys
import time
import urllib.error
import urllib.parse
import urllib.request

BASE = "https://api.figma.com"
MAX_BACKOFF_S = 60  # Figma has returned Retry-After values of days. Cap it.
MAX_RETRIES = 4


def parse_figma_url(url):
    """Node ids arrive hyphenated in URLs and are colon-separated everywhere else."""
    parts = urllib.parse.urlparse(url)
    segments = parts.path.split("/")
    if len(segments) < 3 or not segments[2]:
        raise ValueError(f"no fileKey in {url}")
    raw = urllib.parse.parse_qs(parts.query).get("node-id", [None])[0]
    return {"fileKey": segments[2], "nodeId": raw.replace("-", ":") if raw else None}


class RateLimited(Exception):
    def __init__(self, retry_after_seconds):
        self.retry_after_seconds = retry_after_seconds
        super().__init__(f"rate limited; Figma asked for {retry_after_seconds}s")


class FigmaClient:
    def __init__(self, token=None, token_type="pat"):
        self.token = token or os.environ.get("FIGMA_TOKEN", "")
        if not self.token:
            raise ValueError("set FIGMA_TOKEN")
        self.token_type = token_type

    def _headers(self):
        # Crossing these returns 403, which reads as a file-permission problem.
        if self.token_type == "oauth":
            return {"Authorization": f"Bearer {self.token}"}
        return {"X-Figma-Token": self.token}

    def _get(self, path):
        for attempt in range(MAX_RETRIES + 1):
            req = urllib.request.Request(BASE + path, headers=self._headers())
            try:
                with urllib.request.urlopen(req, timeout=30) as res:
                    return json.loads(res.read())
            except urllib.error.HTTPError as e:
                if e.code == 429:
                    asked = int(e.headers.get("retry-after") or 0)
                    # Past the cap the budget is gone; retrying eats the next window.
                    if asked > MAX_BACKOFF_S or attempt >= MAX_RETRIES:
                        raise RateLimited(asked) from None
                    time.sleep(max(asked, 2**attempt))
                    continue
                if e.code >= 500 and attempt < MAX_RETRIES:
                    time.sleep(2**attempt)
                    continue
                body = e.read().decode("utf-8", "replace")[:300]
                raise RuntimeError(f"{e.code} {path}: {body}") from None

    def me(self):
        """Cheapest call that proves the token works. Run it before anything long."""
        return self._get("/v1/me")

    def file(self, key, depth=None, ids=None):
        """`depth` keeps a large design file from arriving whole."""
        q = {}
        if depth:
            q["depth"] = depth
        if ids:
            q["ids"] = ",".join(ids)
        suffix = f"?{urllib.parse.urlencode(q)}" if q else ""
        return self._get(f"/v1/files/{key}{suffix}")

    def nodes(self, key, ids, depth=None):
        """Specific subtrees. Much smaller than the whole file."""
        q = {"ids": ",".join(ids)}
        if depth:
            q["depth"] = depth
        return self._get(f"/v1/files/{key}/nodes?{urllib.parse.urlencode(q)}")

    def images(self, key, ids, fmt="svg", scale=1):
        """Returns short-lived URLs, not image bytes. Fetch them promptly."""
        q = {"ids": ",".join(ids), "format": fmt, "scale": scale}
        return self._get(f"/v1/images/{key}?{urllib.parse.urlencode(q)}")

    def versions(self, key):
        return self._get(f"/v1/files/{key}/versions")

    def comments(self, key):
        return self._get(f"/v1/files/{key}/comments")


if __name__ == "__main__":
    try:
        client = FigmaClient()
        me = client.me()
        print(json.dumps({"authenticated_as": me.get("handle"), "email": me.get("email")}, indent=2))

        if len(sys.argv) < 2:
            print("pass a Figma URL to read a node")
            sys.exit(0)

        ref = parse_figma_url(sys.argv[1])
        print(json.dumps(ref, indent=2))
        data = (
            client.nodes(ref["fileKey"], [ref["nodeId"]], depth=2)
            if ref["nodeId"]
            else client.file(ref["fileKey"], depth=1)
        )
        print(json.dumps(data, indent=2)[:1500])
    except Exception as exc:
        print(f"FAILED {exc}", file=sys.stderr)
        sys.exit(1)

File: .claude/skills/mcp-api-integration/figma-client.ts
// Figma REST client. Node 18+ (native fetch), no dependencies.
//   FIGMA_TOKEN=<personal access token> npx tsx figma-client.ts
// An OAuth access token works too — pass tokenType: 'oauth'.

const BASE = 'https://api.figma.com';
const MAX_BACKOFF_MS = 60_000; // Figma has returned Retry-After values of days. Cap it.
const MAX_RETRIES = 4;

type TokenType = 'pat' | 'oauth';

export interface FigmaRef {
  fileKey: string;
  nodeId: string | null;
}

/** Node ids arrive hyphenated in URLs and are colon-separated everywhere else. */
export function parseFigmaUrl(url: string): FigmaRef {
  const u = new URL(url);
  const fileKey = u.pathname.split('/')[2];
  if (!fileKey) throw new Error(`no fileKey in ${url}`);
  const raw = u.searchParams.get('node-id');
  return { fileKey, nodeId: raw ? raw.replace(/-/g, ':') : null };
}

export class RateLimited extends Error {
  constructor(public retryAfterSeconds: number) {
    super(`rate limited; Figma asked for ${retryAfterSeconds}s`);
  }
}

export class FigmaClient {
  constructor(
    private token = process.env.FIGMA_TOKEN ?? '',
    private tokenType: TokenType = 'pat',
  ) {
    if (!this.token) throw new Error('set FIGMA_TOKEN');
  }

  private headers(): Record<string, string> {
    // Crossing these returns 403, which reads as a file-permission problem.
    return this.tokenType === 'oauth'
      ? { Authorization: `Bearer ${this.token}` }
      : { 'X-Figma-Token': this.token };
  }

  private async get<T>(path: string): Promise<T> {
    for (let attempt = 0; ; attempt++) {
      const res = await fetch(`${BASE}${path}`, { headers: this.headers() });
      if (res.ok) return (await res.json()) as T;

      if (res.status === 429) {
        const asked = Number(res.headers.get('retry-after') ?? 0);
        // Past the cap the budget is genuinely gone; retrying eats the next window.
        if (asked * 1000 > MAX_BACKOFF_MS || attempt >= MAX_RETRIES) throw new RateLimited(asked);
        await sleep(Math.max(asked * 1000, 2 ** attempt * 1000));
        continue;
      }
      if (res.status >= 500 && attempt < MAX_RETRIES) {
        await sleep(2 ** attempt * 1000);
        continue;
      }
      throw new Error(`${res.status} ${path}: ${(await res.text()).slice(0, 300)}`);
    }
  }

  /** Cheapest call that proves the token works. Run it before anything long. */
  me() {
    return this.get<{ id: string; email: string; handle: string }>('/v1/me');
  }

  /** `depth` keeps a large design file from arriving whole. */
  file(key: string, opts: { depth?: number; ids?: string[] } = {}) {
    const q = new URLSearchParams();
    if (opts.depth) q.set('depth', String(opts.depth));
    if (opts.ids?.length) q.set('ids', opts.ids.join(','));
    return this.get<any>(`/v1/files/${key}${q.size ? `?${q}` : ''}`);
  }

  /** Specific subtrees. Much smaller than the whole file. */
  nodes(key: string, ids: string[], depth?: number) {
    const q = new URLSearchParams({ ids: ids.join(',') });
    if (depth) q.set('depth', String(depth));
    return this.get<any>(`/v1/files/${key}/nodes?${q}`);
  }

  /** Returns short-lived URLs, not image bytes. Fetch them promptly. */
  images(key: string, ids: string[], format: 'png' | 'svg' | 'jpg' | 'pdf' = 'svg', scale = 1) {
    const q = new URLSearchParams({ ids: ids.join(','), format, scale: String(scale) });
    return this.get<{ images: Record<string, string | null>; err: string | null }>(
      `/v1/images/${key}?${q}`,
    );
  }

  versions(key: string) {
    return this.get<any>(`/v1/files/${key}/versions`);
  }

  comments(key: string) {
    return this.get<any>(`/v1/files/${key}/comments`);
  }
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

if (require.main === module) {
  (async () => {
    const client = new FigmaClient();
    const me = await client.me();
    console.log(JSON.stringify({ authenticated_as: me.handle, email: me.email }, null, 2));

    const url = process.argv[2];
    if (!url) return console.log('pass a Figma URL to read a node');

    const { fileKey, nodeId } = parseFigmaUrl(url);
    console.log(JSON.stringify({ fileKey, nodeId }, null, 2));
    const data = nodeId
      ? await client.nodes(fileKey, [nodeId], 2)
      : await client.file(fileKey, { depth: 1 });
    console.log(JSON.stringify(data, null, 2).slice(0, 1500));
  })().catch((e) => {
    console.error('FAILED', e.message);
    process.exit(1);
  });
}
